core/rio/hdt/src/main/java/org/eclipse/rdf4j/rio/hdt/HDTWriter.java

```diff
@@ -0,0 +1,369 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.rio.hdt;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import org.apache.commons.io.output.CountingOutputStream;
+
+import org.eclipse.rdf4j.model.Statement;
+
+import org.eclipse.rdf4j.rio.RDFFormat;
+import org.eclipse.rdf4j.rio.RDFHandlerException;
+import org.eclipse.rdf4j.rio.RioSetting;
+
+import org.eclipse.rdf4j.rio.helpers.AbstractRDFWriter;
+import org.eclipse.rdf4j.rio.helpers.HDTWriterSettings;
+
+/**
+ * <strong>Experimental<strong> RDF writer for HDT v1.0 files. Currently only suitable for input that can fit into
+ * memory.
+ * 
+ * Unfortunately the draft specification is not entirely clear and probably slightly out of date, since the open source
+ * reference implementation HDT-It seems to implement a slightly different version. This parser tries to be compatible
+ * with HDT-It 1.0.
+ * 
+ * The most important parts are the Dictionaries containing the actual values (S, P, O part of a triple), and the
+ * Triples containing the numeric references to construct the triples.
+ * 
+ * Since objects in one triple are often subjects in another triple, these "shared" parts are stored in a shared
+ * Dictionary, which may significantly reduce the file size.
+ * 
+ * File structure:
+ * 
+ * <pre>
+ * +---------------------+
+ * | Global              |
+ * | Header              |
+ * | Dictionary (Shared) |
+ * | Dictionary (S)      |
+ * | Dictionary (P)      |
+ * | Dictionary (O)      |    
+ * | Triples             |
+ * +---------------------+
+ * </pre>
+ * 
+ * @author Bart Hanssens
+ * 
+ * @see <a href="http://www.rdfhdt.org/hdt-binary-format/">HDT draft (2015)</a>
+ * @see <a href="https://www.w3.org/Submission/2011/03/">W3C Member Submission (2011)</a>
+ * 
+ * @since 3.2
+ */
+public class HDTWriter extends AbstractRDFWriter {
+	private final OutputStream out;
+
+	// various counters
+	private long cnt;
+
+	// TODO: rewrite this to cater for larger input
+	// create dictionaries and triples, with some size estimations
+	private static int SIZE = 1_048_576 * 4;
+	private Map<String, Integer> dictShared = new HashMap<>(SIZE / 4);
+	private Map<String, Integer> dictS = new HashMap<>(SIZE / 8);
+	private Map<String, Integer> dictP = new HashMap<>(SIZE / 1024);
+	private Map<String, Integer> dictO = new HashMap<>(SIZE / 2);
+	private List<int[]> tripleRefs = new ArrayList<>(SIZE);
```

Comment:
I gather these lists and maps are the main reason this writer currently only works for input that fits in memory?  If so it might be worth looking at using MapDB to overflow to disk - we use it in several other places for this purpose, and it works quite well.

Reply:
Also: I'm not sufficiently familiar with HDT to understand how this works fully, but I assume these lists are kept to allow compression (using a short reference if a particular value reoccurs instead of serializing the entire value again)? If so, you could take a leaf out of the BinaryRDFWriter's approach, which has a configurable buffer for this kind of thing: it keeps references for compression in a fifo queue so that often re-occuring values can be referenced in shorthand, but it allows for dropping references from the queue and re-creating a full value (with a _new_ reference hook) later on in the file.

Reply:
Thanks, very useful ! I'll look into it.

Indeed, HDT
- first replaces the parts  of triples with numerical references, so it has to keep track of what string is already assigned a number
- writes out some statistics on the total number of triples, different subjects etc
- then orders the string (well, bytes actually) representation of these parts and compresses them into "blocks"
- renumbers the numerical references and encodes them as a sort of layers
- writes out these compressed blocks
- and only then writes out the encoded references

Reply:
Hm, if it requires that you write out statistics at the start of the file, and also requires a full mapping dictionary at the start that can not be changed later on in the file, I'm not sure how you would ever implement this in a streaming fashion.

Reply:
Maybe if you split it into three files and then merge them later?

Reply:
Full streaming is indeed not possible AFAIK....

I've already left out a few statistics that didn't seem to be used by HDT-It, and maybe I can leave out a few others...
But the "commonly" used encoding still seem to require e.g. total number of triples somewhat upfront, some parts need the encoded length... and there is also the sorting/renumbering of all triples.
AFAIK the HDT format does not have "chunks" (i.e. to encode per X million triples)

Reply:
Benefit though is that we can write a read-only sail wrapper that works with an in-place HDT file (ExtensibleSail to the rescue). We could also look into writing a `dump` tool for the NativeStore that dumps the data in HDT format, which should be considerably more performant since i believe the NativeStore is already sorted (B-tree).

Reply:
Indeed that is my long-term plan :-)

HDT is meant for read-only, fast lookup for linked data fragments / triple fragments applications

Reply:
I've noticed the MapDB in RDF4J is quite old (1.0.8) and not supported anymore. 
Newer versions have more sophisticated settings (e.g. overflow to disk), is there a specific reason to sticking to 1.0.8 ?

Reply:
Nothing beyond "we haven't had the time or resources to upgrade", I think. I vaguely remember there may have been some non-compatible changes that makes upgrading non-trivial. But I haven't looked at it in a while.

Reply:
If it has built-in support for overflow now that would be good - I believe we currently implement our own overflow by juggling normal collections and then at some point switching to MapDB.